crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -0,0 +1,568 @@
+//! Applies the protocol fee to the solution received from the solver.
+//!
+//! Solvers respond differently for the sell and buy orders.
+//!
+//! EXAMPLES:
+//!
+//! SELL ORDER
+//! Selling 1 WETH for at least `x` amount of USDC. Solvers respond with
+//! Fee = 0.05 WETH
+//! Executed = 0.95 WETH
+//!
+//! This response is adjusted by the protocol fee of 0.1 WETH:
+//! Fee = 0.05 WETH + 0.1 WETH = 0.15 WETH
+//! Executed = 0.95 WETH - 0.1 WETH = 0.85 WETH
+//!
+//! BUY ORDER
+//! Buying 1 WETH for at most `x` amount of USDC. Solvers respond with
+//! Fee = 0.05 WETH
+//! Executed = 1 WETH
+//!
+//! This response is adjusted by the protocol fee of 0.1 WETH:
+//! Fee = 0.05 WETH + 0.1 WETH = 0.15 WETH
+//! Executed = 1 WETH
+
+use {
+    super::trade::{Fee, Fulfillment, InvalidExecutedAmount},
+    crate::domain::{
+        competition::{
+            order,
+            order::{FeePolicy, Side},
+        },
+        eth,
+    },
+};
+
+impl Fulfillment {
+    /// Applies the protocol fee to the existing fulfillment creating a new one.
+    pub fn with_protocol_fee(&self, prices: ClearingPrices) -> Result<Self, Error> {
+        let protocol_fee = self.protocol_fee(prices)?;
+
+        // Increase the fee by the protocol fee
+        let fee = match self.raw_fee() {
+            Fee::Static => Fee::Static,
+            Fee::Dynamic(fee) => Fee::Dynamic((fee.0 + protocol_fee).into()),
+        };
+
+        // Reduce the executed amount by the protocol fee. This is because solvers are
+        // unaware of the protocol fee that driver introduces and they only account
+        // for their own fee.
+        let order = self.order().clone();
+        let executed = match order.side {
+            order::Side::Buy => self.executed(),
+            order::Side::Sell => order::TargetAmount(
+                self.executed()
+                    .0
+                    .checked_sub(protocol_fee)
+                    .ok_or(Error::Overflow)?,
+            ),
+        };
+
+        Fulfillment::new(order, executed, fee).map_err(Into::into)
+    }
+
+    fn protocol_fee(&self, prices: ClearingPrices) -> Result<eth::U256, Error> {
+        let mut protocol_fee = eth::U256::zero();
+        for fee_policy in self.order().fee_policies.iter() {
```

Comment:
Probably not an issue for now, but I wonder if the current logic would work for multiple fee policies (e.g. two 50% price improvements). I would think those would be applied sequentially (ie. we first get a fulfillment from applying the first fee - with less surplus! - and then only take 50% of the remaining surplus rather than 100%).

My suggestion would be to expose another `apply` method on a local impl extension of the FeePolicy enum, that takes a fulfilment and prices and returns a new fulfilment with its policy applied.

But we can also cross that bridge when we get to it, in this case we should probably document and guard with an error that multiple fee policies aren't yet properly implemented.

Reply:
I need to think more if we should allow for multiple policies of the same kind. Does it make sense to have two price improvement policies on the same order, where the calculated fee amounts depend on the ordering of policies.
Would you expect two policies, one price improvement, one volume based, to also be sequentially applied?